src/go/k8s/api/redpanda/v1alpha2/user_types.go

```diff
@@ -0,0 +1,193 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Defines the desired state of the Redpanda user.
+	Spec UserSpec `json:"spec"`
+	// Represents the current status of the Redpanda user.
+	// +kubebuilder:default={conditions: {{type: "Synced", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status UserStatus `json:"status,omitempty"`
+}
+
+var (
+	_ KafkaConnectedObject     = (*User)(nil)
+	_ AdminConnectedObject     = (*User)(nil)
+	_ ClusterReferencingObject = (*User)(nil)
+)
+
+// RedpandaName identifies the unique username created within a Redpanda cluster
+// for a specific user based on its namespace and name in Kubernetes.
+func (u *User) RedpandaName() string {
+	if u.Spec.UsernameOverride != nil {
+		return *u.Spec.UsernameOverride
+	}
+	return fmt.Sprintf("%s/%s", u.Namespace, u.Name)
+}
+
+// ACLName constructs the name of a User for defining ACLs.
+func (u *User) ACLName() string {
+	return "User:" + u.RedpandaName()
+}
+
+func (u *User) GetKafkaAPISpec() *KafkaAPISpec {
+	return u.Spec.KafkaAPISpec
+}
+
+func (u *User) GetAdminAPISpec() *AdminAPISpec {
+	return u.Spec.AdminAPISpec
+}
+
+func (u *User) GetClusterRef() *ClusterRef {
+	return u.Spec.ClusterRef
+}
+
+// UserSpec defines the configuration of a Redpanda user.
+// +kubebuilder:validation:XValidation:message="either clusterRef or kafkaApiSpec and adminApiSpec must be set",rule="has(self.clusterRef) || (has(self.kafkaApiSpec) && has(self.adminApiSpec))"
+type UserSpec struct {
+	// ClusterRef is a reference to the cluster where the user should be created.
+	// It is used in constructing the client created to configure a cluster.
+	// This takes precedence over KafkaAPISpec and AdminAPISpec.
+	ClusterRef *ClusterRef `json:"clusterRef,omitempty"`
+	// KafkaAPISpec is the configuration information for communicating with the Kafka
+	// API of a Redpanda cluster where the User should be created.
+	KafkaAPISpec *KafkaAPISpec `json:"kafkaApiSpec,omitempty"`
+	// AdminAPISpec is the configuration information for communicating with the Admin
+	// API of a Redpanda cluster where the User should be created.
+	AdminAPISpec *AdminAPISpec `json:"adminApiSpec,omitempty"`
+	// Authentication defines the authentication information for a user. If no
+	// Authentication credentials are specified, then no user will be created.
+	// This is useful when wanting to manage ACLs for an already-existing user.
+	Authentication *UserAuthenticationSpec `json:"authentication,omitempty"`
+	// Authorization rules defined for this user.
+	Authorization *UserAuthorizationSpec `json:"authorization,omitempty"`
+	// Template to specify how user secrets are generated.
+	Template *UserTemplateSpec `json:"template,omitempty"`
+	// UsernameOverride allows for specifying a particular username for the
+	// given user rather than the calculated "<NAMESPACE>/<NAME>".
+	UsernameOverride *string `json:"usernameOverride,omitempty"`
```

Comment:
Thoughts on this -- if we don't want to allow for cross-namespace ClusterRef eventually (and don't mind the fact that you could still break yourself with competing `User` definitions via specifying a `KafkaAPISpec` for a user in a different namespace (assuming we eventually unscope the operator from only managing a single namespace), then we could just drop the `<NAMESPACE>/<NAME>` convention and assume that we can just create/manage each user based on the object `Name` and drop this override field.

Reply:
@david-yu Is my handle. I'm not sure I fully understand the question here but sounds like may need to ask the customer whether they envision running the operator to manage multiple redpanda clusters across multiple namespaces?

Reply:
So @david-yu the main thing is that our current controller is entirely namespace-scoped (i.e. it can't even read resources created in other namespaces), hence, if you deploy the controller to `redpanda-system`, you need to create all your custom resources in `redpanda-system`, which, will in turn, create all your clusters in `redpanda-system`.

We have previously talked about how we don't particularly like the namespace-scoping of the controller and generally feel it should be cluster-scoped, but there hasn't been any firm decision on that. However, if we allow cluster-scoping you could envision doing things like:

- creating a cluster in the `foo` namespace
- creating a user `a` in the `bar` namespace that references the `foo` namespace cluster
- creating a user `a` in the `baz` namespace that references the `foo` namespace cluster

Now you have two resources, one in the `bar`, one in the `baz` namespace attempting to both manage and overwrite each other's configurations for user `a`, that is unless you name the users in Redpanda `bar/a` and `baz/a` respectively.

The two workarounds are basically what I mentioned above:
1. Either allow for someone to specify a username directly as an override so someone can manually avoid conflicts (or cause them and break themselves)
2. Only allow for "local object references", i.e. a user can only be associated with a cluster in its own namespace -- this works so long as you're using a `ClusterRef`, but would potentially break if someone is manually defining connection parameters with `Kafka/AdminApiSpec` fields.

Reply:
I think I'll just go ahead and assume only local object refs for now.

Reply:
I don't know enough yet to provide guidance, but sounds like you have considered the options available and are making the best judgement.